clarify credential error response can be authorization or credential request error (#64)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

When the Credential Request is invalid or unauthorized, the Credential Issuer constructs the error response as defined in this section.

#### Authorization Errors {#authorization-errors}

> If the Credential Request does not contain an Access Token that enables issuance of a requested credential, the Credential Endpoint MUST return an error response as defined in section 3 of [@!RFC6750].

Reviewer: ```suggestion
If the Credential Request does not contain an Access Token that enables issuance of a requested credential, the Credential Endpoint returns an authorization error response such as defined in section 3 of [@!RFC6750].
```

> If the Credential Request does not contain an Access Token that enables issuance of a requested credential, the Credential Endpoint MUST return an authorization error response as defined in section 3 of [@!RFC6750].

Reviewer: again 
```suggestion
If the Credential Request does not contain an Access Token that enables issuance of a requested credential, the Credential Endpoint returns an authorization error response such as defined in section 3 of [@!RFC6750].
```

> If the Credential Request does not contain an Access Token that enables issuance of a requested credential, the Credential Endpoint returns an authorization error response such as defined in section 3 of [@!RFC6750].

Reviewer 2: ```suggestion
If the Credential Request does not contain an Access Token that enables the issuance of a requested credential, the Credential Endpoint MUST use the HTTP status code 401 (Unauthorized) with the content type set to `application/json` and the error code set to `invalid_token`, as defined in section 3 of [@!RFC6750].
```

Author: `invalid_token` error code should not be mandated. and it is not application/json. and 401 is already covered by `as defined by rfc6750`.

```suggestion
If the Credential Request does not contain an Access Token that enables issuance of a requested credential, the Credential Endpoint MUST return an authorization error response as defined in section 3 of [@!RFC6750] .
```

Reviewer: > invalid_token error code should not be mandated. and it is not application/json. and 401 is already covered by as defined by rfc6750.

+1 

but changing it to have that MUST would preclude challenging for dpop bound access tokens. The text in question here shouldn't change.

Reviewer: please put this back to "If the Credential Request does not contain an Access Token that enables issuance of a requested credential, the Credential Endpoint returns an authorization error response such as defined in section 3 of [@!RFC6750]."

> For the errors specific to the Credential Request such as those caused by `type`, `format`, `proof` or encryption parameters in the request, error codes parameters defined in (#credential-request-errors) SHOULD be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].

Reviewer: ```suggestion
For the errors specific to the Credential Request such as those caused by `type`, `format`, `proof`, or encryption parameters in the request, error codes parameters defined in (#credential-request-errors) SHOULD be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].
```

> For the errors specific to the Credential Request such as those caused by `type`, `format`, `proof`, or encryption parameters in the request, error codes parameters defined in (#credential-request-errors) SHOULD be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].

Reviewer 2: ```suggestion
If the Credential Request produces an error caused by the values ​​of the `type`, `format`, `proof` members or caused by the encryption parameters in the request, the error codes values defined in (#credential-request-errors) SHOULD be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].
```

> For the errors specific to the Credential Request such as those caused by `type`, `format`, `proof`, or encryption parameters in the request, the error codes values defined in (#credential-request-errors) SHOULD be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].

Reviewer 3: Wouldn't a MUST be appropriate?

Author: i do not see a reason why not..

Author: ```suggestion
For the errors specific to the Credential Request such as those caused by `type`, `format`, `proof`, or encryption parameters in the request, the error codes values defined in (#credential-request-errors) MUST be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].
```

> For the errors specific to the Credential Request such as those caused by `type`, `format`, `proof`, or encryption parameters in the request, the error codes values defined in (#credential-request-errors) SHOULD be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].

Reviewer 3: I find this paragraph underneath #### Authorization Errors {#authorization-errors} misleading. Suggest to remove the section heading and just have two paragraphs or to move the second section into the credential issuance errors section.

Author: I think it would be more misleading to have two separate paragraphs talking about errors for credential endpoint. we can change section titles (credential error response; authorization errors; credential request errors) if they are misleading, but I really do not think we need to change the structure.

Author: discussed in the WG call, agreed to move line 889 under 891.

#### Credential Request Errors {#credential-request-errors}

> If the Credential Request is requesting for the issuance of a credential not supported by the Credential Endpoint, the HTTP response body uses the `application/json` media type with an HTTP 400 (Bad Request) status code (unless specified otherwise) and includes the following parameters with the response:

Reviewer 2: ```suggestion
If the Credential Request requires issuing a credential that is not supported by the Credential Endpoint, the HTTP response MUST use the HTTP status code 400 (Bad Request) and set the content type to `application/json`, with the following parameters in the response body:
```

Author: ```suggestion
If the Wallet is requesting the issuance of a credential that is not supported by the Credential Endpoint, the HTTP response MUST use the HTTP status code 400 (Bad Request) and set the content type to `application/json` with the following parameters in the response body::
```

* `error`: REQUIRED. A key at the top level of a JSON object, the value of which SHOULD be a single ASCII [@!USASCII] error code from the following:
  * `invalid_credential_request`: The Credential Request is missing a required parameter, includes an unsupported parameter or parameter value, repeats the same parameter, or is otherwise malformed.

>   * `unsupported_credential_type`: Requested credential type is not supported.

Reviewer 4: `type` is specific to the `*_vc_*` (W3C) formats, right? So, probably it shouldn't be in these format-independent `error` values.

Reviewer 5: A more general error could be `unsupported_credential`. This could be used to target non format related credential errors of this klnd.

Author: type is not specific to W3C formats :) it is used to refer to a claimset to be included in a credential. the definition of doctype in ISO 18013-5 is `docType is the requested document type.` and we use the following definition of doctype in VCI too
> doctype: REQUIRED. JSON string identifying the credential type as defined in [[ISO.18013-5](https://openid.github.io/OpenID4VCI/openid-4-verifiable-credential-issuance-1_0.html#ISO.18013-5)].

so I think the current text is ok

Reviewer 3: That's a slippery slope as the spec treats any credential type information as format specific. In addition, W3C has an array of type values. I assume the issuer would like to indicate which of them is unsupported.

Author: how should your comment be interpreted? are you agreeing with Jacob and Pedro that a generic use of a term `type` in the spec should be revisited? i do not disagree but that should be a separate issue.

  * `unsupported_credential_format`: Requested credential format is not supported.

>   * `invalid_proof`: The `proof` in the Credential Request was invalid. For example, no key proof was provided (the `proof` field was not present); the provided key proof was not bound to a nonce provided by the Credential Issuer, etc.

Reviewer 2: ```suggestion
  * `invalid_proof`: The `proof` in the Credential Request is invalid. The `proof` field is not present or the provided key proof is invalid or not bound to a nonce provided by the Credential Issuer.
```

  * `invalid_encryption_parameters`: This error occurs when the encryption parameters in the Credential Request are either invalid or missing. In the latter case, it indicates that the Credential Issuer requires the Credential Response to be sent encrypted, but the Credential Request does not contain the necessary encryption parameters.

> * `error_description`: OPTIONAL. A key at the top level of a JSON object, the value of which MUST be a human-readable ASCII [@!USASCII] text providing additional information, used to assist the client developer in understanding the error that occurred. Values for the `error_description` parameter MUST NOT include characters outside the set %x20-21 / %x23-5B / %x5D-7E.

Reviewer 2: ```suggestion
* `error_description`: OPTIONAL. A key at the top level of a JSON object, which value MUST be an human-readable ASCII [@!USASCII] text, providing any additional information used to assist the client implementers in understanding the occurred error. The values for the `error_description` parameter MUST NOT include characters outside the set `%x20-21 / %x23-5B / %x5D-7E`.
```

Author: ```suggestion
* `error_description`: OPTIONAL. A key at the top level of a JSON object, whose value MUST be a human-readable ASCII [@!USASCII] text, providing any additional information used to assist the client implementers in understanding the occurred error. The values for the `error_description` parameter MUST NOT include characters outside the set `%x20-21 / %x23-5B / %x5D-7E`.
```

The usage of these parameters takes precedence over the `invalid_request` parameter defined in (#authorization-errors), since they provide more details about the errors.

> The following is a non-normative example of a Credential Error Response an unsupported credential format was requested:

Reviewer: ```suggestion
The following is a non-normative example of a Credential Error Response where an unsupported credential format was requested:
```

When the Credential Request is invalid or unauthorized, the Credential Issuer constructs the error response as defined in this section.

#### Authorization Errors {#authorization-errors}

If the Credential Request does not contain an Access Token that enables issuance of a requested Credential, the Credential Endpoint returns an authorization error response such as defined in section 3 of [@!RFC6750].

#### Credential Request Errors {#credential-request-errors}

> For the errors specific to the Credential Request such as those caused by `type`, `format`, `proof`, or encryption parameters in the request, the error codes values defined in this section MUST be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].

Author: ```suggestion
For the errors specific to the payload of the Credential Request such as those caused by `type`, `format`, `proof`, or encryption parameters in the request, the error codes values defined in this section MUST be used instead of a generic `invalid_request` parameter defined in section 3.1 of [@!RFC6750].
```

## Deferred Credential Error Response {#deferred-credential_error_response}

> When the Deferred Credential Request is invalid or the credential is not available yet, the Credential Issuer constructs the error response as defined in (#credential-error-response).

Reviewer 2: ```suggestion
When the Deferred Credential Request is invalid or the credential is not available yet, the Credential Issuer SHOULD use the error response as defined in (#credential-error-response).
```

Author: why SHOULD? this feels like a MUST, also for consistency
```suggestion
When the Deferred Credential Request is invalid or the credential is not available yet, the Credential Issuer MUST construct the error response as defined in (#credential-error-response).
```

> When the Deferred Credential Request is invalid or the credential is not available yet, the Credential Issuer MUST construct the error response as defined in (#credential-error-response).

Author: ```suggestion
When the Deferred Credential Request is invalid or the credential is not available yet, the Credential Issuer constructs the error response as defined in (#credential-error-response).
```

> When the Deferred Credential Request is invalid or unauthorized, or the credential is not available yet, the Credential Issuer constructs the error response as defined in (#credential-error-response).

Reviewer: ```suggestion
When the Deferred Credential Request is invalid or the credential is not available yet, the Credential Issuer constructs the error response as defined in (#credential-error-response).
```

The following additional error codes are specified in addition to those already defined in (#credential-request-errors):

> * `issuance_pending` - The credential issuance is still pending. The error response will also contain another claim `interval` determining the minimum amount of time in seconds that the Wallet SHOULD wait between requests to the Deferred Credential Endpoint.  If no value is provided, clients MUST use 5 as the default.

Reviewer 2: ```suggestion
* `issuance_pending` - The credential issuance is still pending. The error response SHOULD also contain the `interval` member, determining the minimum amount of time in seconds that the Wallet SHOULD wait before providing a new request to the Deferred Credential Endpoint.  If `interval` member is missing or its value is not provided, the clients MUST use 5 as the default value.
```

Author: ```suggestion
* `issuance_pending` - The credential issuance is still pending. The error response SHOULD also contain the `interval` member, determining the minimum amount of time in seconds that the Wallet needs to wait before providing a new request to the Deferred Credential Endpoint.  If `interval` member is missing or its value is not provided, the Wallet MUST use `5` as the default value.
```

> * `invalid_transaction_id` - Deferred Credential Request contained an invalid `transaction_id`, i.e., it was not issued by the respective Credential Issuer or was already used to obtain the Credential.

Reviewer 2: ```suggestion
* `invalid_transaction_id` - The Deferred Credential Request contains an invalid `transaction_id`. The `transaction_id` was not issued by the respective Credential Issuer or it was already used to obtain the Credential.
```

> * `issuance_pending` - The credential issuance is still pending. The error response SHOULD also contain the `interval` member, determining the minimum amount of time in seconds that the Wallet needs to wait before providing a new request to the Deferred Credential Endpoint.  If `interval` member is missing or its value is not provided, the Wallet MUST use `5` as the default value.

Reviewer 2: ```suggestion
* `issuance_pending` - The Credential issuance is still pending. The error response SHOULD also contain the `interval` member, determining the minimum amount of time in seconds that the Wallet needs to wait before providing a new request to the Deferred Credential Endpoint.  If `interval` member is missing or its value is not provided, the Wallet MUST use `5` as the default value.
```
